### PR TITLE
Feature/views

### PR DESCRIPTION
**Pull request by iamsmkr**

### CODE REVIEW

File: docbrown/core/src/tgraph.rs [hunk 1]
```diff
@@ -50,22 +50,24 @@ impl TemporalGraph {
             .unwrap_or(0)
     }
 
-    pub(crate) fn has_edge(&self, src: u64, dst: u64) -> bool {
-        let v_pid = self.logical_to_physical[&src];
+    pub(crate) fn has_edge(&self, v1: u64, v2: u64) -> bool {
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:53]: This is undirected? I think we need to change it to directed

Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:53]: I think we also need to have a windowed version of this no?

File: docbrown/core/src/tgraph.rs [hunk 2]
```diff
@@ -301,54 +294,169 @@ impl TemporalGraph {
...
-            },
+            Adj::Solo(lid) => VertexView { g_id: lid, pid },
+            Adj::List { logical, .. } => VertexView { g_id: logical, pid },
         });
 
         Box::new(vs)
     }
 
-    pub(crate) fn neighbours(
+    pub(crate) fn edge(&self, v1: u64, v2: u64) -> Option<EdgeView> {
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:328]: As above, I think what you have created here is almost the undirected version. If the user asks for 1->2 we shouldn't return 2->1

File: docbrown/core/src/tgraph.rs [hunk 3]
```diff
@@ -301,54 +294,169 @@ impl TemporalGraph {
...
+                            e_meta: e,
+                        }),
+                        _ => None,
+                    }
+                }
+            }
+        }
+    }
+
+    pub(crate) fn edge_window(&self, v1: u64, v2: u64, w: &Range<i64>) -> Option<EdgeView> {
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:386]: I believe the same here as above.

File: docbrown/core/src/tgraph.rs [hunk 4]
```diff
@@ -301,54 +294,169 @@ impl TemporalGraph {
...
+                        }
+                    }
+                }
+            },
+            false => Option::<EdgeView>::None,
+        }
+    }
+
+    // FIXME: all the functions using global ID need to be changed to use the physical ID instead
+    pub(crate) fn edges(
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:455]: This probably needs to be called something like vertex_edges to make it a bit clearer its for a singular node.

File: docbrown/core/src/tgraph.rs [hunk 5]
```diff
@@ -357,102 +465,337 @@ impl TemporalGraph {
...
+        let edges = self.edges(v, d);
+
+        Box::new(edges.map(move |edge| {
+            let EdgeView { src_id, dst_id, .. } = edge;
+
+            let src_g_id = edge.src_g_id;
+            let dst_g_id = edge.dst_g_id;
+
+            if v == src_g_id {
+                VertexView::new(dst_g_id, dst_id)
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:591]: This fine even if the vertex isn't on the same shard?

File: docbrown/core/src/tgraph.rs [hunk 6]
```diff
@@ -357,102 +465,337 @@ impl TemporalGraph {
...
+            } else {
+                VertexView::new(src_g_id, src_id)
+            }
+        }))
+    }
+
+    pub(crate) fn neighbours_ids(
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:623]: Do we actually use any of these global (non-windowed functions anywhere. I.e. would it be easier to have everything have a window and remove?

File: docbrown/core/src/tgraph.rs [hunk 7]
```diff
@@ -1064,85 +1248,44 @@ mod graph_test {
...
-
-        g.add_vertex(1, 9);
-        g.add_edge_remote_into(2, 7, 9, &vec![("bla".to_string(), Prop::U32(1))]);
-
-        let actual: bool = g.has_edge(9, 7);
-        assert_eq!(actual, true);
+        assert_eq!(g.has_edge(8, 9), true);
+        assert_eq!(g.has_edge(9, 8), true);
+        assert_eq!(g.has_edge(9, 11), true);
+        assert_eq!(g.has_edge(11, 9), true);
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:1286]: From the comment above, this should return false, the edge 11,9 was never added.

File: docbrown/core/src/tgraph.rs [hunk 8]
```diff
@@ -1595,9 +1748,9 @@ mod graph_test {
         assert_eq!(edge_weights_out_11, vec![(&4, 1122.0), (&4, 1133.0)]);
 
         let edge_weights_into_11 = g
-            .neighbours(11, Direction::IN)
+            .edges(11, Direction::IN)
             .flat_map(|e| {
-                e.props("weight").map(|i| {
+                g.edge_prop(e.e_meta.edge_meta_id(), "weight").map(|i| {
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph.rs:1753]: This probably could be renamed I think

File: docbrown/core/src/tgraph_shard.rs [hunk 9]
```diff
@@ -179,100 +156,213 @@ impl TGraphShard {
...
+            for v_id in iter {
+                yield_!(v_id)
             }
         });
 
-        Box::new(iter.into_iter())
+        iter.into_iter()
+    }
+
+    pub fn edges_window_t(
```
Comment (miratepuffin) [inline docbrown/core/src/tgraph_shard.rs:241]: At all levels if this function could be renamed to make it a bit clearer what is the different between it and edges_window - as they have such similar method signatures as well.

File: docbrown/db/src/graph.rs [hunk 10]
```diff
@@ -116,8 +119,8 @@ impl Graph {
         self.shards.iter().map(|shard| shard.out_edges_len()).sum()
     }
 
-    pub fn has_edge(&self, src: u64, dst: u64) -> bool {
-        self.shards.iter().any(|shard| shard.has_edge(src, dst))
+    pub fn has_edge(&self, v1: u64, v2: u64) -> bool {
```
Comment (miratepuffin) [inline docbrown/db/src/graph.rs:122]: I think we need to benchmark this as it seems like we are going to be doing a linear search every time we look up, making algos which use it O(n^2)

File: docbrown/db/src/graph.rs [hunk 11]
```diff
@@ -333,6 +410,39 @@ mod db_tests {
         fs::remove_dir_all(tmp_docbrown_path).unwrap();
     }
 
+    #[test]
+    fn has_edge() {
+        let g = Graph::new(2);
+        g.add_edge(1, 7, 8, &vec![]);
+
+        assert_eq!(g.has_edge(8, 7), true);
```
Comment (miratepuffin) [inline docbrown/db/src/graph.rs:418]: This should be false, 7->8 was added.

File: pyraphtory/src/graph_window.rs [hunk 12]
```diff
@@ -73,21 +102,95 @@ impl WindowedVertex {
...
+
+#[pyclass]
+pub struct WindowedEdge {
+    pub edge_id: usize,
+    #[pyo3(get)]
+    pub src: u64,
+    #[pyo3(get)]
+    pub dst: u64,
+    #[pyo3(get)]
+    pub t: Option<i64>,
```
Comment (miratepuffin) [inline pyraphtory/src/graph_window.rs:168]: We probably need to rename this

File: pyraphtory/src/graph_window.rs [hunk 13]
```diff
@@ -73,21 +102,95 @@ impl WindowedVertex {
...
+    #[pyo3(get)]
+    pub is_remote: bool,
```
Comment (miratepuffin) [inline pyraphtory/src/graph_window.rs:170]: Is there any need for us to expose this in python

**Review by miratepuffin — APPROVED** [review]
Absolutely outstanding work @iamsmkr . Some comments throughout, but more for us to discuss tomorrow in the office. 

Only two major things are:

-  I think we need to rename functions like `edge_window` and `edge_window_t` to make it clearer what they are doing - hard to disambiguate for the user.
- Your `has_edge` function is undirected at the moment, this should be a simple fix though.
- Is there really any reason for us to have the non windowed versions of any functions when everything is a window.

File: docbrown/core/src/tgraph.rs (hunk 5 above)
Comment (iamsmkr) [reply docbrown/core/src/tgraph.rs:591]: I'll write a test for this!

File: pyraphtory/src/graph_window.rs (hunk 13 above)
Comment (iamsmkr) [reply pyraphtory/src/graph_window.rs:170]: no

**Pull request merged**